### DISCUSSION TITLE
Webhooks consistency + slug handling + semantic revisions

### DISCUSSION THREAD

**Discussion by luciandavidescu** [directus/directus#10981] [Feature Requests]
_Here are these 3 issues that I think are somehow linked to one another, so I put them in a single post. This came out quite long, so you may want to skip towards the end for a quick summary of the title._

First of all, I can't help to start by saying how enthusiastic I am about this BRILLIANT piece of software – it's the CMS I've been dreaming of for the last 10 years and I just wish I had discovered it much sooner... well, it's sort of like the bittersweet feeling of finding true love at old age.

I've been exploring it for the last week and I'm amazed at how it has everything it should but not one thing above that. (Strapi is a close call but rather unintuitive and, well... one wouldn't dare go to a newsroom and tell them "here's your new CMS", Ghost chose to stay too opinionated and restrictive and more along those lines).

I'm not a professional developer, only a journalist who had to do 'my own carpentry' because nothing was quite fit to purpose. Add in a penchant for minimalism and a bit of OCD and you'll see where I'm coming from.

So now that I have the worthy tool, I started to build the frontend I've been planning for a while (the functional logic may seem peculiar at first, but bear with me). 

It's a static site generator relying on webhooks and ~DB queries (could just as well be API calls, but I chose to do DB because I'm more comfortable with that right now)~ GraphQL API queries. 

My Express.js app will listen to a port for webhooks sent by the CMS, query the database accordingly, use a templating engine of choice (going with Handlebars most likely) and then do the CRUD on the filesystem. 

The purpose is to combine the power of a DB driven CMS with the speed of purely static pages, while also achieving better stability, security and ease of deployment, without the need of a complex and inherently unstable filecache.

The basic concept demonstrator [[edit+edit: here it is, rewritten for GraphQL]](https://github.com/luciandavidescu/directus-static-frontend/blob/main/directus-static-frontend.js) is just 30ish lines of code that do a basic 'website', with homepage and articles. Lots to do from here, when it's all finished, I'll release it as open source.

### My use case
The default and only webhook schema is workable, but not in the most straightforward way. I guess it makes perfect sense for the backend, but not so much for the frontend. That might still be good enough, if not for the fear of future breaking changes. So it would be great if either a) custom schemas were possible or b) there's an alternative default schema that's relevant to the frontend, which I'll detail a bit lower.

Now, I'll concede that my use case is quite narrow, but I'm sure there may be others that could use this. I'm also aware that things could be done in another, simpler, way that I just haven't figured out yet, so if that's the case kindly point it to me.

### My app's logic
First of all, I need to identify the updated item by ID. There are 3(!) different places it could be in: `key` OR `keys[0]` OR `event:items.delete` > `payload`. (foreaches for batch updates make things even more complicated, but enough for now).

Then, I have one of four scenarios that need be fulfilled:

**1. Do nothing**
The item is not yet supposed to be public, so just break if, with the event: 'items.create', the status is 'archived' OR null ('draft').

**2. Delete**
A file has to be deleted in one of 3 cases – event:'items.delete' OR event:'items.update' + status.draft OR event: 'items.update' + status.archived. (LE: When an item is deleted completely, its slug is only retrievable from past revisions)

**3. Create or update**
A file is created on event:items create OR is overwritten on event:items.update. (LE. Right now there's no solid way to tell if an item was updated as `draft` or as `published` without asking the DB first)

**4. Create and delete**
Sometimes the url will be changed, so there are more things to be done. When `'payload'` contains `'slug'`, do the usual from point 3, but also delete the file from the old location (and add a proper redirect). So I'll also have to find my way through _directus-revisions data /delta_. Not quite straightforward, didn't thoroughly explore this yet, I suppose it can be done... or else either write my own DB field each time (expensive, less than secure due to having to grant write access) or keep a separate record (still expensive, less than reliable by adding complexity).

### I. Webhooks consistency
The schema – either selectable or customizable - that I see working nicely for a frontend environment would include these 3 items:
- `'key'` (or `'keys'`), constant in all webhooks (so even on deletes, just like 'event' or 'accountability').
- `'visibility'`, that could be simply OFF (for drafts, archives, deletes) or ON (for published, updated). LE. Or simply 'status' always sent.
- `'delta'+'slug'`, so it saves an expensive (maybe even unreliable?) trip to the database or api.

### II. Slug revision handling
Now, I'm aware that the last item would be the most complicated an contetious. 

First, i suppose that diffs in the payload would be no easy feat. Second, I started with exploring the demo for Directus 8, so I'm aware that the former 'slug' field ceased to have its own life and melted into a generic feature set. This is useful and abides to the philosophy that makes Directus so great.

However, I feel it's worth pointing out that there's a more generic semantic that the slug (or uri, or permalink, or pretty whatever) belongs to, and that is the _public identifier_ (in the loose sense, not necessarily according to whatever specification). Like for example with SSN vs name: they are strictly linked, the latter may change, but that is a special event that needs be formally notified, for proper record keeping. Or like with all concepts encompassed by terms like alias, moniker, aka, or – since we're here – the latin _cognomen_. 

Even if they may change once in a while, they do remain part of the subject's identity and may even come back to life at any point (like in _the Artist Formerly Known as..._) or preserve their core meaning (_What's in a name? That which we call a rose, by any other name..._).

### III. Semantic revisions

In DB language, they'd be not mere transient revisions, but semantic fields holding an array of one or more previously used values.

So, ok, not the 'slug field' again, but... what if there was this checkable option to shadow a field or another and store all previous revisions in their own column?

Beyond the obvious uri/"slug", I could think of some other generic use-cases, like:
- User management: A user may change their name, but the app needs to be able to match old references (including those in plain text or maybe hardcoded) to the acutual ID. Or they may have multiple e-mail addresses as usernames to login with etc.
- Security: Store previous hash values to check if there's an attempt of password reuse.
- Mergers, acquisitions and spin-offs: Whether it's a financial stock quote, or a software fork/unification that involves rebranding, it's being able to follow data throughout.
- Product codes: A company may want to change the code schema of some of all of their products, but still be able to match by the more familiar former version.
- Car registration/insurance: Be able to check changing license plates against constant chassis number.
- [LE: For one reason or another (more readable accounability, backwards compatibility etc.) a user may want to store the full previous versions of a certain fields, not just the diffs]

### Summing up questions:

- **I.** What to expect for the future? ('won't fix' / non-breaking changes / relevant features coming / dunno right now).
- **II.** What may be the safest way to do get the revisioned item with what's available now?
- **III.** Can we have that pleeeeease?

**Comment by rijkvanzanten** [comment]
> First of all, I can't help to start by saying how enthusiastic I am about this BRILLIANT piece of software – it's the CMS I've been dreaming of for the last 10 years and I just wish I had discovered it much sooner... well, it's sort of like the bittersweet feeling of finding true love at old age.

Thanks! Means a lot ❤️ 

> Add in a penchant for minimalism and a bit of OCD and you'll see where I'm coming from. 

Hehe, both @benhaynes and myself do fully understand what that's like 👨🏻‍🎨 

> The purpose is to combine the power of a DB driven CMS with the speed of purely static pages, while also achieving better stability, security and ease of deployment, without the need of a complex and inherently unstable filecache.

Makes sense 👍🏻 

> i. What to expect for the future? ('won't fix' / non-breaking changes / relevant features coming / dunno right now).

Non-breaking changes for starters. More importantly: we're currently planning an enhancement to webhooks that will allow for _way_ more flexible configuration in what's sent to the endpoint, and when it's sent. For example, that will allow you to say "only send the webhook when field X = Y" and what fields are sent exactly in what format. This should help out a ton when it comes to sending submitting consistent webhooks at expected times.

As for the key vs keys inconsistently, I totally agree. We should submit the payload in the same format where possible (cc @nickrum, should be easy enough to add an additional `keys` next to the existing `key` that's just `keys: [key]` for consistency's sake).

That being said, the above in-works improvement will also cover this by simply allowing the user to choose what keys to submit, and what they're called.

> ii. What may be the safest way to do get the revisioned item with what's available now?

[The /revisions endpoint](https://docs.directus.io/reference/system/revisions/) holds the delta and snapshots of the items at every point they're changed. You should be able to fetch the current / previous revisions of any given item through that endpoint, to reconstruct a history of all changes to a given field.

> iii. Can we have that pleeeeease?

Interesting! How would it store multiple old revisions at the same time? Relational databases[^1] aren't the best suited at storing list values, so you'd almost have to have a separate table where every row holds a single revision of the linked column 🤔 Do you have an example of how you might structure such a thing?

**Reply by luciandavidescu** [reply]
> Non-breaking changes for starters. More importantly: we're currently planning an enhancement to webhooks that will allow for way more flexible configuration (...) the above in-works improvement will also cover this by simply allowing the user to choose what keys to submit, and what they're called.

That pretty much does it 

> The /revisions endpoint holds the delta and snapshots

Ah, swift, will get from there, I was hasty going with DB i guess. (LE. Indeed, switching to API, saves a lot of trouble and the speed cost is quite insignificant for my use case)

> How would it store multiple old revisions at the same time? (...) Do you have an example of how you might structure such a thing?

Basically would be a push to array (or unshift even better) on change event, for the specified field only, so something like this:

column: slug
value: current-perfect-slug-until-we-change-it-again

column: slug_aka
value: ["former-perfect-slug-that-people-linked-to","former-slug-with-a-spelling-error-that-we-only-saw-after-some-time","first-attempt-at-a-slug-that-was-not-so-perfect-but-got-shared-on-social-media"]

I for one woult then assign this to a tags type field in the backend, so some of them could be "released" manually, at will.

**Reply by rijkvanzanten** [reply]
> Basically would be a push to array (or unshift even better) on change event

The issue with that is that not all database vendors support proper searching through nested json array values 🤔 so it'll prevent things like searching through previously used values

**Reply by luciandavidescu** [reply]
Ah, I'm not too familiar with DB logic, never thought of that. But it would still be just as helpful if you already knew exactly where to look for, as in this context, right?

**Reply by rijkvanzanten** [reply]
Yes and no. For seeing the previous values of the same field, it would work just fine, as you're already operating on the same row. But if you want to take it to the next step with things like "Make sure the new value hasn't been used in the past" (like your password example from earlier), you'd run into a problem, as there's no good way to search through all used values at that point 🤔

**Reply by luciandavidescu** [reply]
but... I only really care if _this_ user is attempting to reuse a password, not if they happened to stumble upon a previous password of some _other_ user. (LE – indeed, it does make some of the use cases difficult to properly implement if db engines are not always full-in, so I get the point)

**Reply by luciandavidescu** [reply]
otoh, even a single 'previous' might be good enough for +90% of use cases :D
(LE: while for the other 10% users might just chain the process, creating as many 'aka' columns as needed.)

**Reply by luciandavidescu** [reply]
> As for the key vs keys inconsistently, I totally agree. We should submit the payload in the same format where possible (cc @nickrum, should be easy enough to add an additional keys next to the existing key that's just keys: [key] for consistency's sake).

Maybe 'status' might also be always sent, as I now realise that it's nearly impossible to tell if an item was updated as draft or as published.

**Reply by joselcvarela** [reply]
> > As for the key vs keys inconsistently, I totally agree. We should submit the payload in the same format where possible (cc @nickrum, should be easy enough to add an additional keys next to the existing key that's just keys: [key] for consistency's sake).
> 
> Maybe 'status' might also be always sent, as I now realise that it's nearly impossible to tell if an item was updated as draft or as published.

The `status` column is specific to your particular table, like `articles`. But we need to be more abstract, so you can access to properties of the document via `payload` passed as argument when available, otherwise you would need to fetch the item from right collection and filter on your side.

**Reply by luciandavidescu** [reply]
@joselcvarela 

The issue here is that  status is not in the payload unless changed, so the client would have to check DB with every 'items.update'; with people editing their drafts much more often then their published pieces, that's plenty of unnecessary trips. While the underlying abstract concept is that visibility status is of critical significance for the frontend, unlike most others.

But I agree, this is much too specific. Ideally, we could pick exactly what to send in the payload... and then API/DB access might not even be necessary at all(!)

Anyway, by now I finished the basic logic that I was detailing above (<del>bar for</del><ins>including</ins> batch updates) and indeed everything can be made to work somehow; some things are not that straightforward (ended up with two separate delete events), but it does get the job done.

https://github.com/luciandavidescu/directus-static-frontend/blob/main/directus-static-frontend.js

**Reply by joselcvarela** [reply]
I think the _Project Waterway_ will be suitable for this use-case, am I right @rijkvanzanten ?

**Reply by luciandavidescu** [reply]
Is this _Project Waterway_ that keeps coming up lately related to this?

> we're currently planning an enhancement to webhooks that will allow for way more flexible configuration in what's sent to the endpoint, and when it's sent. For example, that will allow you to say "only send the webhook when field X = Y" and what fields are sent exactly in what format.

Any timeframe for it?

**Reply by benhaynes** [reply]
Haha, as indicated by the "Project" codename, we're not quite at a point where we can discuss this feature yet... but are hoping to have something public in the next few weeks.

**Reply by luciandavidescu** [reply]
Taking a wild guess... deliver content with no API/DB callbacks whatsoever (much better reliability, security and simplicity). Maybe even relay it to multiple independent users' apps, with full usable data and completely agnostic to origin. Thus do push notifications, newsletters, transactional e-mails, standalone browser or phone apps, backup services, countless others some of which are yet to be thought of. That should be ABOVE BRILLIANT!
Do you sell stock?

**Comment by luciandavidescu** [comment]
On the UI, it may be something like this...

![fields](https://user-images.githubusercontent.com/26859276/149083028-5cc07ae5-99fe-47eb-abd6-fd8e4f24be65.png)

...and also with an option to actually name or pick the field. Or if this doesn't make sense, maybe it could be done with a configurable rule.

**Comment by luciandavidescu** [comment]
@rijkvanzanten 
> The /revisions endpoint holds the delta and snapshots of the items at every point they're changed. You should be able to fetch the current / previous revisions of any given item through that endpoint, to reconstruct a history of all changes to a given field.

Not easier to do (just like in the DB ,`delta` mirrors `data` items on the respective row so need to dig further), but it did help me understand how things work.

So, I can just get the last but one revisions' `data` with something like:

```
query {
	revisions(    
	filter: {
	item: {
		_eq: "MY_ID"
		}          
	}
	sort:["-activity"]
	limit:1
	offset:1
	)  
	{
	data
	}
}
```

Then pull out my previous slug with:

```
fetch(endpoint, payload).then(response => response.json()).then(response => {
    let aka = response['data']['revisions'][0]['data']['slug']
    console.log(aka)
})
// (I suppose there must be a way to retrieve just the single item in the first place, nevermind, I'll figure that out)
```

Hopefully, this stays reliable in all possible scenarios.

Taking as little of your time as possible – just tell me if this is the optimal approach or else kindly help point me to the piece in the code that does this thing:

![revision](https://user-images.githubusercontent.com/26859276/149541010-defd0cda-a562-454a-99bf-4f24540e611e.png)

**Reply by joselcvarela** [reply]
You can also use CSV field to hold previous data with custom API hooks 
https://docs.directus.io/extensions/hooks/#custom-api-hooks

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
